app/code/Magento/Quote/Model/Cart/AddProductsToCart.php

```diff
@@ -0,0 +1,222 @@
+        $this->requestBuilder = $requestBuilder;
+    }
+
+    /**
+     * Add cart items to the cart
+     *
+     * @param string $maskedCartId
+     * @param Data\CartItem[] $cartItems
+     * @return AddProductsToCartOutput
+     * @throws NoSuchEntityException Could not find a Cart with provided $maskedCartId
+     */
+    public function execute(string $maskedCartId, array $cartItems): AddProductsToCartOutput
+    {
+        $cartId = $this->maskedQuoteIdToQuoteId->execute($maskedCartId);
+        $cart = $this->cartRepository->get($cartId);
+
+        foreach ($cartItems as $n => $cartItem) {
+            $this->addItemToCart($cart, $cartItem, $n);
+        }
+        if ($cart->getData('has_error')) {
+            $errors = $cart->getErrors();
+
+            /** @var MessageInterface $error */
+            foreach ($errors as $error) {
+                $this->addError($error->getText());
+            }
+        }
+        $this->cartRepository->save($cart);
+
+        return $this->prepareErrorOutput($cart);
+    }
+
+    /**
+     * Adds a particular item to the shopping cart
+     *
+     * @param CartInterface|Quote $cart
+     * @param Data\CartItem $cartItem
+     * @param int $cartItemPosition
+     */
+    private function addItemToCart(CartInterface $cart, Data\CartItem $cartItem, int $cartItemPosition): void
+    {
+        $sku = $cartItem->getSku();
+
+        try {
+            $product = $this->productRepository->get($sku, false, null, true);
+        } catch (NoSuchEntityException $e) {
+            $this->addError(
+                __('Could not find a product with SKU "%sku"', ['sku' => $sku])->render(),
+                $cartItemPosition,
+                self::ERROR_PRODUCT_NOT_FOUND
+            );
+
+            return;
+        }
+
+        try {
+            $result = $cart->addProduct($product, $this->requestBuilder->build($cartItem));
+        } catch (\Throwable $e) {
+            $this->addError(
+                __(
+                    'Could not add the product with SKU %sku to the shopping cart: %message',
+                    ['sku' => $sku, 'message' => $e->getMessage()]
+                )->render(),
+                $cartItemPosition
+            );
+            return;
+        }
+
+        if (is_string($result)) {
+            $errors = array_unique(explode("\n", $result));
+            foreach ($errors as $error) {
+                $this->addError(__($error)->render(), $cartItemPosition);
+            }
+        }
+    }
+
+    /**
+     * Add order line item error
+     *
+     * @param string $message
+     * @param int $cartItemPosition
+     * @param string|null $code
+     * @return void
+     */
+    private function addError(string $message, int $cartItemPosition = 0, string $code = ''): void
+    {
+        $this->errors[] = new Data\Error(
+            $message,
+            $code ?? $this->getErrorCode($message),
+            $cartItemPosition
```

Comment:
Also I found a related issue. when we keep adding products to the cart, and if we run into errors like out of stock, the cart becomes empty. 
![Jul-10-2020 00-15-10](https://user-images.githubusercontent.com/43612864/87119075-92845e80-c242-11ea-80a0-39a6376c6d2c.gif)

Reply:
That's a good catch. The problem is it's a default system behavior introduced to the core a couple of months ago.

https://github.com/magento/magento2/blob/faced716b53f4d152113d0a0171cfc171ecbefa0/app/code/Magento/Quote/Model/Quote.php#L1684

I'm not sure about the best way to overcome it.

Reply:
Hey @rogyar. This took me a while to figure it out. Apparently this is happening only on the new mutation, not the existing ones. 

Even though the cart item is being deleted in core, the cart is not saved. But I see in this PR the cart is being saved here even when there are errors. app/code/Magento/Quote/Model/Cart/AddProductsToCart.php:116. I don't see any to reason to do that, please correct me if I am wrong.

So fixing that should fix this issue.

Reply:
Hi @prabhuram93. That's a good point. But unfortunately, without explicit cart saving on `Model/Cart/AddProductsToCart.php:116` the quote is not being saved anywhere. You will still get the correct cart output since the corresponding model with the current changes will be used. But you will not get the record saved in the database.

You may test it in the following way: 
- Set stock qty of some product (let's say "simple1") to 10
- Comment out the following line `Model/Cart/AddProductsToCart.php:116`
- Add "simple1" to the shopping cart with QTY 3
- Add "simple1" using the same mutation (QTY 3)
- Check the result, you will have total QTY 3 (not 6) because the result of the very first mutation is not being saved anywhere. 

So I'm still thinking that the approach of removing item if it has an error
https://github.com/magento/magento2/blob/faced716b53f4d152113d0a0171cfc171ecbefa0/app/code/Magento/Quote/Model/Quote.php#L1684

Does the dirty thing. 
I'm going to spend more time to overcome this weird behavior but any other idea will be very appreciated. 

Thank you!

Reply:
Hey @rogyar. I understand that cart is not being saved anywhere else. But I don't see any reason to save the cart when there are errors. So in the scenario you mentioned I wouldn't just comment out the line in step 2, rather we can handle it to save the cart when only there are no errors. That way when there are errors just the error objects will be updated for that particular mutation and the cart will not be saved. When there are no errors cart will be saved. I managed to try it out and it works for both the cases you mentioned. Let me know if this is still unclear. Thanks!

Reply:
Got it, thank you. We may adjust it. The only thing that will be lost in this case, described in the following scenario.

- Add 3 different items to the cart
- There's no issue with first 2 but 3rd is problematic
- We will have no items in the cart instead of having 2 of them

... but, it's another story :)

Reply:
Hey @rogyar, I don't see a use case where we add 3 different items to the cart at the same point. We might do it one after the other, but not together. So we are good on that front.